2024/08/migration-from-struts1-to-struts2.md

```diff
@@ -0,0 +1,174 @@
+
+#### Request handling
+
+Configure Struts 2 to handle requests at `*.action` extension where Struts 1 handles at `*.do` extension.
+
+#### Configure web application
+
+Web application is enabled in `web.xml` configuration file. There are a few changes needed for Struts 2 in this file. Dispatcher configuration is just done by renaming the configuration tags from `<servlet>` and `<servlet-mapping>` to `<filter>` and `<filter-mapping>`.
+
+A very basic `web.xml` configuration for Struts 1 would be like below to show the settings.
+
+```xml
+<servlet>
+    <servlet-name>action</servlet-name>
+    <servlet-class>org.apache.struts.action.ActionServlet</servlet-class>
+    <init-param>
+        <param-name>config</param-name>
+        <param-value>/WEB-INF/struts-config.xml</param-value>
+    </init-param>
+    <load-on-startup>2</load-on-startup>
+</servlet>
+
+<servlet-mapping>
+    <servlet-name>action</servlet-name>
+    <url-pattern>*.do</url-pattern>
+</servlet-mapping>
+```
+
+The corresponding conversion to Struts 2 be like:
+
+```xml
+<filter>
+    <filter-name>webwork</filter-name>
+    <filter-class>org.apache.struts.action2.dispatcher.FilterDispatcher</filter-class>
+</filter>
+
+<filter-mapping>
+    <filter-name>webwork</filter-name>
+    <url-pattern>/*</url-pattern>
+</filter-mapping>
+```
+
+#### Migrating actions
+
+A general structure of a Struts 1 action class is like below.
+
+```java
+public class CustomAction extends Action {
+    public ActionForward execute(
+        ActionMapping mapping,
+        ActionForm form,
+        HttpServletRequest request,
+        HttpServletResponse response) throws Exception {
+        // add logic
+        return (mapping.findForward("success"));
+    }
+}
+```
+
+Struts 1 actions are singleton and extends base `Action` class. Alternatively `DispatchAction` can be used but the entrry point to an Action class is `execute` method for both implementation. Actions has to be thread-safe and hence all the needed variables are handled in method scope. Finally an action methods return `ActionForward` response.
+
+On the other hand, a Struts 2 a basic action class structure would be:
+
+```java
+public class MyAction extends ActionSupport {
+   public String execute() throws Exception {
+        // add logic
+        return "success";
+   }
+}
+```
+
+A Struts 2 action class not necessarily but usually extends `ActionSupport` and usual entry point is again `execute()` method without any parameters. In Struts 2 action classes are not singleton and creates instance for each request, so Class scope variables can be used. Struts 2 finally injects HttpServletRequest into ServletRequestAware actions to complete the request flow.
+
+#### Action Configuration
+
+In Struts 1 applications action configurations are made in `struts-config.xml` located in `WEB-INF` directory. Struts 1 applications need to configure actions and action form beans in this configuration file.
```

Comment:
In Struts 1 applications, action configurations are made in `struts-config.xml` located in the `WEB-INF` directory. Struts 1 applications need to configure actions and action form beans in this configuration file.